File: frappe/creche_module/report/daily_creche_attendance_activity/daily_creche_attendance_activity.py
import frappe
import calendar
from frappe.utils import nowdate
from datetime import datetime, timedelta, date

def execute(filters=None):
    columns = get_columns(filters)
    data = get_summary_data(filters)
    return columns, data

def get_columns(filters):
    columns = [
        {"label": "Partner", "fieldname": "partner", "fieldtype": "Data", "width": 120},
        {"label": "State", "fieldname": "state", "fieldtype": "Data", "width": 120},
        {"label": "District", "fieldname": "district", "fieldtype": "Data", "width": 120},
        {"label": "Block", "fieldname": "block", "fieldtype": "Data", "width": 120},
        {"label": "Gram Panchayat", "fieldname": "gp", "fieldtype": "Data", "width": 150},
        {"label": "Creche", "fieldname": "creche_name", "fieldtype": "Data", "width": 160},
        {"label": "Creche ID", "fieldname": "creche_id", "fieldtype": "Data", "width": 150},
        {"label": "Creche IDX", "fieldname": "creche_idx", "fieldtype": "Data", "width": 150},
        {"label": "Creche Opening Date", "fieldname": "creche_opening_date", "fieldtype": "Data", "width": 170},
        {"label": "Total Days Submitted", "fieldname": "submitted", "fieldtype": "Int", "width": 200},
        {"label": "Total Days Not Submitted", "fieldname": "not_submitted", "fieldtype": "Int", "width": 200},
    ]
    
    month = int(filters.get("month", nowdate().split('-')[1]))
    year = int(filters.get("year", nowdate().split('-')[0]))
    last_day = calendar.monthrange(year, month)[1]

    for day in range(1, last_day + 1):
        columns.append({
            "label": f"{day:02d}-{month:02d}-{year}",
            "fieldname": f"day_{day}",
            "fieldtype": "Data",
            "width": 120
        })
        
    return columns

@frappe.whitelist()
def get_summary_data(filters=None):
    partner = frappe.db.get_value("User", frappe.session.user, "partner")
    month = int(filters.get("month", nowdate().split('-')[1]))
    year = int(filters.get("year", nowdate().split('-')[0]))
    last_day = calendar.monthrange(year, month)[1]

    # filters logic for cr_opening starts here
    cstart_date, cend_date = None, None
    range_type = filters.get("cr_opening_range_type") if filters.get("cr_opening_range_type") else None

    state_query = """ 
        SELECT DISTINCT ts.name AS state_id
        FROM `tabState` ts 
        JOIN `tabUser Geography Mapping` ugm ON ugm.state_id = ts.name
        WHERE ugm.parent = %s 
        ORDER BY ts.state_name
    """
    state_params = (frappe.session.user,)
    current_user_state = frappe.db.sql(state_query, state_params, as_dict=True)
    state = filters.get("state") or (current_user_state[0]['state_id'] if current_user_state else None)

    state = None if not state else state

    if range_type:
        single_date = filters.get("single_date")
        date_range = filters.get("c_opening_range")

        if single_date and isinstance(single_date, str):
            single_date = datetime.strptime(single_date, "%Y-%m-%d").date()
            
        if range_type == "between" and date_range and len(date_range) == 2:
            cstart_date, cend_date = date_range

        elif range_type == "before" and single_date:
            cstart_date, cend_date = date(2017, 1, 1), single_date - timedelta(days=1)

        elif range_type == "after" and single_date:
            cstart_date, cend_date = single_date + timedelta(days=1), date.today()

        elif range_type == "equal" and single_date:
            cstart_date = cend_date = single_date

    # filters logic for cr_opening ends here
    creche_status_id = filters.get("creche_status_id") if filters else None
    phases_cleaned = ",".join(p.strip() for p in filters["phases"].split(",") if p.strip().isdigit()) if filters.get("phases") else None

    filters_dict = {
        'selected_partner': partner if partner else filters.get('partner'),
        'state': state,
        'district': filters.get('district'),
        'supervisor_id': filters.get('supervisor_id'),
        'block': filters.get('block'),
        'gp': filters.get('gp'),
        'creche': filters.get('creche'),
        'month': month,
        'year': year,
        'cstart_date': cstart_date,
        'cend_date': cend_date,
        'creche_status_id': creche_status_id,
        'phases': phases_cleaned
    }

    conditions = []
    conditions_att = []

    if filters_dict['selected_partner']:
        conditions.append("cr.partner_id = %(selected_partner)s")
        conditions_att.append("tca.partner_id = %(selected_partner)s")
    if filters_dict['state']:
        conditions.append("cr.state_id = %(state)s")
        conditions_att.append("tca.state_id = %(state)s")
    if filters_dict['district']:
        conditions.append("cr.district_id = %(district)s")
        conditions_att.append("tca.district_id = %(district)s")
    if filters_dict['block']:
        conditions.append("cr.block_id = %(block)s")
        conditions_att.append("tca.block_id = %(block)s")
    if filters_dict['gp']:
        conditions.append("cr.gp_id = %(gp)s")
        conditions_att.append("tca.gp_id = %(gp)s")
    if filters_dict['creche']:
        conditions.append("cr.name = %(creche)s")
        conditions_att.append("tca.creche_id = %(creche)s")
    if filters_dict['supervisor_id']:
        conditions.append("cr.supervisor_id = %(supervisor_id)s")
    if cstart_date and cend_date:
        conditions.append("cr.creche_opening_date BETWEEN %(cstart_date)s AND %(cend_date)s")
    if filters_dict['phases']:
        conditions.append("(%(phases)s IS NULL OR FIND_IN_SET(cr.phase, %(phases)s))")

    if filters_dict['creche_status_id']:
        conditions.append("(%(creche_status_id)s IS NULL OR cr.creche_status_id = %(creche_status_id)s)")


    condition_str = " AND ".join(conditions) if conditions else "1=1"
    conditions_att_str = " AND ".join(conditions_att) if conditions_att else "1=1"

    
    current_date = date.today()

    # daily_columns = [
    #     f"IFNULL(catt.day_{day}, IF(DATE('{year}-{month:02d}-{day:02d}') > DATE('{current_date}'), 'N/A', 'N')) AS day_{day}"
    #     for day in range(1, last_day + 1)
    # ]
    # daily_columns_pc = [f"MAX(CASE WHEN DAY(date_of_attendance) = {day} AND is_shishu_ghar_is_closed_for_the_day = 1 THEN 'N/A' WHEN DAY(date_of_attendance) = {day} THEN 'Y' END) AS day_{day}" for day in range(1, last_day + 1)]

    daily_columns = [
        f"""
        CASE
            WHEN cr.creche_opening_date IS NULL THEN 'N/A'
            WHEN DATE('{year}-{month:02d}-{day:02d}') < cr.creche_opening_date THEN 'N/A'
            WHEN DATE('{year}-{month:02d}-{day:02d}') > DATE('{current_date}') THEN 'N/A'
            WHEN catt.day_{day} IS NULL THEN 'Not Submitted'
            WHEN catt.day_{day} = 'N/A' THEN 'Closed'
            ELSE catt.day_{day}
        END AS day_{day}
        """
        for day in range(1, last_day + 1)
    ]

    daily_columns_pc = [
        f"""
        MAX(
            CASE
                WHEN DAY(date_of_attendance) = {day} AND is_shishu_ghar_is_closed_for_the_day = 1 THEN 'Closed'
                WHEN DAY(date_of_attendance) = {day} THEN 'Open'
            END
        ) AS day_{day}
        """
        for day in range(1, last_day + 1)
    ]
    
    sql_query = f"""
    SELECT
        p.partner_name AS partner,
        s.state_name AS state,
        d.district_name AS district,
        b.block_name AS block,
        g.gp_name AS gp,
        cr.creche_name AS creche_name,
        cr.creche_id AS creche_id,
        cr.name AS creche_idx,
        DATE_FORMAT(cr.creche_opening_date, '%%d-%%m-%%Y') AS 'creche_opening_date',
        {", ".join(daily_columns)}
    FROM `tabCreche` cr
    LEFT JOIN (
        SELECT creche_id, {", ".join(daily_columns_pc)}
        FROM `tabChild Attendance` tca
        WHERE YEAR(date_of_attendance) = {year} 
        AND MONTH(date_of_attendance) = {month} 
        AND {conditions_att_str}
        GROUP BY creche_id
    ) AS catt ON catt.creche_id = cr.name
    JOIN `tabState` s ON cr.state_id = s.name
    JOIN `tabDistrict` d ON cr.district_id = d.name
    JOIN `tabBlock` b ON cr.block_id = b.name
    JOIN `tabGram Panchayat` g ON cr.gp_id = g.name
    JOIN `tabVillage` v ON cr.village_id = v.name
    JOIN `tabPartner` p ON cr.partner_id = p.name
    WHERE {condition_str}
    GROUP BY p.name, s.name, d.name, b.name, g.name, v.name, cr.name
    ORDER BY p.partner_name, s.state_name, d.district_name, b.block_name, g.gp_name, cr.creche_name
    """

    data = frappe.db.sql(sql_query, filters_dict, as_dict=True)
    for row in data:
        submitted_count = sum(1 for day in range(1, last_day + 1) if row.get(f"day_{day}") == "Open" or row.get(f"day_{day}") == "Closed") 
        not_submitted_count = sum(1 for day in range(1, last_day + 1) if row.get(f"day_{day}") == "Not Submitted")
        row["submitted"] = submitted_count
        row["not_submitted"] = not_submitted_count

    daily_totals = {}
    for day in range(1, last_day + 1):
        daily_totals[f"day_{day}"] = sum(1 for row in data if row.get(f"day_{day}") in ["Open", "Closed"])

    summary_row = {
        "creche_name": "<b style='color:black;'>Total</b>",
        "submitted": sum(row.get("submitted", 0) for row in data),
        "not_submitted": sum(row.get("not_submitted", 0) for row in data),
    }
    for day in range(1, last_day + 1):
        summary_row[f"day_{day}"] = daily_totals[f"day_{day}"]
    data.append(summary_row)
    return data

File: frappe/creche_module/report/daily_creche_attendance_activity/daily_creche_attendance_activity.js
frappe.query_reports["Daily Creche Attendance Activity"] = {
    filters: [
        {
            fieldname: "year",
            label: __("Year"),
            fieldtype: "Select",
            options: [
                "2024",
                "2025"
            ],
            default: frappe.datetime.get_today().split('-')[0],
        },
        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: [
                { "value": "01", "label": "January" },
                { "value": "02", "label": "February" },
                { "value": "03", "label": "March" },
                { "value": "04", "label": "April" },
                { "value": "05", "label": "May" },
                { "value": "06", "label": "June" },
                { "value": "07", "label": "July" },
                { "value": "08", "label": "August" },
                { "value": "09", "label": "September" },
                { "value": "10", "label": "October" },
                { "value": "11", "label": "November" },
                { "value": "12", "label": "December" }
            ],
            default: frappe.datetime.get_today().split('-')[1],
        },
        {
            fieldname: "partner",
            label: __("Partner"),
            fieldtype: "Link",
            options: "Partner"
        },
		{
			"fieldname": "state",
			"label": __("State"),
			"fieldtype": "Link",
			"options": "State",
			"get_query": function () {
				return {
					filters: {
						"is_active": 1
					}
				};
			},
			"on_change": function () {
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "district",
			"label": __("District"),
			"fieldtype": "Link",
			"options": "District",
			get_query: function () {
				let state = frappe.query_report.get_filter_value("state")
				return state ? { filters: { state_id: state } } : {};
			},
			"on_change": function () {
				frappe.query_report.set_filter_value("block", "");
				frappe.query_report.set_filter_value("gp", "");
				frappe.query_report.set_filter_value("creche", "");
				frappe.query_report.set_filter_value("supervisor_id", "");
				frappe.query_report.refresh();
			}
		},
        {
            fieldname: "block",
            label: __("Block"),
            fieldtype: "Link",
            options: "Block",
        },
        {
            fieldname: "gp",
            label: __("Gram Panchayat"),
            fieldtype: "Link",
            options: "Gram Panchayat",
        },
        {
            fieldname: "creche",
            label: __("Creche"),
            fieldtype: "Link",
            options: "Creche",
            reqd: 0,
            get_query: function () {
                const partner = frappe.query_report.get_filter_value("partner");
                if (partner) {
                    return {
                        filters: {
                            partner_id: partner
                        }
                    };
                } else {
                    return {};
                }
            },
        },
        {
            fieldname: "supervisor_id",
            label: __("Supervisor"),
            fieldtype: "Link",
            options: "User",
            get_query: function () {
                let creche = frappe.query_report.get_filter_value("creche");
                return creche ? { filters: { creche: creche } } : {};
            },
        },
        {
            "fieldname": "phases",
            "label": __("Phase"),
            "fieldtype": "MultiSelect",
            "options": ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10"],
            "reqd": 0,
            "default": ""
        },
        {
            "fieldname": "creche_status_id",
            "label": __("Creche Status"),
            "fieldtype": "Select",
            "options": [
                { "value": "", "label": __("") },
                { "value": "1", "label": __("Planned") },
                { "value": "2", "label": __("Plan dropped") },
                { "value": "3", "label": __("Active/Operational") },
                { "value": "4", "label": __("Closed") },
            ],
            "default": "3",
        },
        {
            fieldname: "cr_opening_range_type",
            label: __("Creche Opening Date"),
            fieldtype: "Select",
            options: [
                { value: "", label: __("") },
                { value: "between", label: __("Between") },
                { value: "before", label: __("Before") },
                { value: "after", label: __("After") },
                { value: "equal", label: __("Equal") }
            ],
            // default: "",
            on_change: function () {
                const dateRangeType = frappe.query_report.get_filter_value("cr_opening_range_type");

                const isBetween = dateRangeType === "between";
                const isSingleDate = ["before", "after", "equal"].includes(dateRangeType);
                const isCleared = dateRangeType === "";

                frappe.query_report.get_filter("c_opening_range").toggle(isBetween);
                frappe.query_report.get_filter("single_date").toggle(isSingleDate);

                if (isBetween) {
                    frappe.query_report.set_filter_value("single_date", "");
                } else if (isSingleDate) {
                    frappe.query_report.set_filter_value("c_opening_range", []);
                }

                if (isCleared) {
                    frappe.query_report.set_filter_value("c_opening_range", []);
                    frappe.query_report.set_filter_value("single_date", "");
                }

                frappe.query_report.refresh();
            }
        },
        {
            fieldname: "c_opening_range",
            label: __("Creche Opening Range"),
            fieldtype: "DateRange",
            hidden: 1
        },
        {
            fieldname: "single_date",
            label: __("Creche Opening Date"),
            fieldtype: "Date",
            hidden: 1
        }
    ],

    formatter: function (value, row, column, data, default_formatter) {


        if (value === null || value === undefined) return "";
        value = default_formatter(value, row, column, data);

        if (value === "Not Submitted") {
            return `<b style='color:#ff6456;'>${value}</b>`;
        }

        if (data.creche_name && data.creche_name.includes("Total")) {
            return `<b style='color:black;'>${value}</b>`;
        }

        let day_match = column.fieldname.match(/day_(\d+)/);
        let date_parts = column.label.split("-");
        let day = day_match ? day_match[1] : null;
        let month = date_parts[1];
        let year = date_parts[2];

        if (value === "Closed" && day && month && year) {
            let element_id = `${data.creche_idx}_${day}_${month}_${year}`;

            return `<span 
            id="${element_id}" 
            title=""
            style="cursor: pointer;"
            onmouseover="fetch_reason_on_hover(this, '${data.creche_idx}', '${year}-${month}-${day}')"
            onmouseleave="hide_tooltip()"
        >${value}</span>`;
        }

        return value;
    },
    onload: function () {
        console.log("Report Loaded: Daily Creche Attendance Activity");

        const style = document.createElement("style");
        style.innerHTML = `
            .dt-instance-1 .dt-cell--col-8  {
                display: none !important;
            }
        `;
        document.head.appendChild(style);
    }
}

const reasonMapping = {
    "1": "Local Festival",
    "2": "Public Holiday",
    "3": "Training",
    "4": "Any Other",
    "5": "Weekly Off"
};

window.fetch_reason_on_hover = function (element, creche_id, date_of_attendance) {
    if (!element.getAttribute("data-fetched")) {
        // show_tooltip(element, "Fetching reason...");

        frappe.call({
            method: "frappe.client.get_value",
            args: {
                doctype: "Child Attendance",
                filters: { creche_id, date_of_attendance },
                fieldname: ["reason_for_closure_id", "reason_other"] // Fetch both fields
            },
            callback: function (response) {
                let reasonId = response.message?.reason_for_closure_id || "No reason provided";
                let otherReason = response.message?.reason_other || "";

                let reason = reasonMapping[reasonId] || "Unknown Reason";

                // If reason is "Any Other", append the other reason
                if (reasonId === "4" && otherReason) {
                    reason = `Any Other - ${otherReason}`;
                }

                element.setAttribute("data-fetched", "true");
                show_tooltip(element, reason);
            },
            error: function () {
                show_tooltip(element, "Failed to fetch reason");
            }
        });
    } else {
        show_tooltip(element, element.getAttribute("data-tooltip") || "No reason provided");
    }
};



function show_tooltip(element, message) {
    let existingTooltip = document.querySelector(".custom-tooltip");
    if (existingTooltip) existingTooltip.remove();

    let tooltip = document.createElement("div");
    tooltip.className = "custom-tooltip";
    tooltip.innerText = message;

    document.body.appendChild(tooltip);

    let rect = element.getBoundingClientRect();
    tooltip.style.position = "absolute";
    tooltip.style.top = `${rect.bottom + window.scrollY + 5}px`;
    tooltip.style.left = `${rect.left + window.scrollX}px`;
    tooltip.style.background = "#333";
    tooltip.style.color = "#fff";
    tooltip.style.padding = "5px 10px";
    tooltip.style.borderRadius = "5px";
    tooltip.style.zIndex = "9999";
    tooltip.style.boxShadow = "0 2px 5px rgba(0,0,0,0.3)";
    tooltip.style.whiteSpace = "nowrap";
    tooltip.style.animation = "fade-in 0.2s ease-in-out";

    element.tooltipElement = tooltip;
    element.setAttribute("data-tooltip", message);
}

window.hide_tooltip = function () {
    let tooltip = document.querySelector(".custom-tooltip");
    if (tooltip) tooltip.remove();
};

